File: crates/periskop-static-scanner/fixtures/java/evasion/RuntimeBuiltUrl.java
package com.example.summaries;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

/**
 * Known gap: the destination is assembled at runtime.
 *
 * The scanner sees a variable where the endpoint should be. The host comes from
 * the environment and the path is concatenated onto it, so no literal in this
 * file names a provider. Reporting it anyway would be a claim the evidence does
 * not support, so it is not reported at all.
 *
 * Catalogued as KG-002 in the known gaps list.
 */
public final class RuntimeBuiltUrl {

    private final HttpClient http = HttpClient.newHttpClient();

    public String summarize(String record) throws Exception {
        String endpoint = System.getenv("MODEL_HOST") + "/v1/chat/completions";

        HttpRequest request = HttpRequest.newBuilder()
                .uri(new URI(endpoint))
                .POST(HttpRequest.BodyPublishers.ofString(record))
                .build();

        return http.send(request, HttpResponse.BodyHandlers.ofString()).body();
    }
}

File: crates/periskop-static-scanner/fixtures/go/evasion/runtime_built_url.go
// Known gap: the destination is assembled at runtime.
//
// The scanner sees an identifier where the URL should be. Reporting it as a
// provider call would be a claim the source does not support, and reporting every
// http.Post with a variable argument would drown the findings that are supported.
//
// The rule that would otherwise catch this file keys on a literal URL, so the
// gap is a property of the evidence rather than of the pattern: no tightening of
// the query recovers a string that is only built while the program runs.
//
// Catalogued as KG-002 in the known gaps list.
package evasion

import (
	"net/http"
	"os"
	"strings"
)

func summarize(record string) (*http.Response, error) {
	endpoint := strings.TrimSuffix(os.Getenv("MODEL_ENDPOINT"), "/") + "/v1/chat/completions"
	return http.Post(endpoint, "application/json", strings.NewReader(record))
}
